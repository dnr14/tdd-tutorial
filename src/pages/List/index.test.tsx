import { screen, render, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { List } from './index';
import { createMemoryHistory } from 'history';
import { Router, useLocation } from 'react-router-dom';
const _todo_key = 'todo_list';

beforeEach(() => {
  localStorage.clear();
  const _list = ['todo1', 'todo2', 'todo3'];
  localStorage.setItem(_todo_key, JSON.stringify(_list));
});

const TestComponent = () => {
  const { pathname } = useLocation();
  return <div>{pathname}</div>;
};

const HOC = ({ path = '/' }: { path?: string }) => {
  const history = createMemoryHistory();

  history.push(path);
  return (
    <Router history={history}>
      <TestComponent />
      <List></List>
    </Router>
  );
};

describe('<List>', () => {
  it('<List> 랜더링 확인', () => {
    const history = createMemoryHistory();
    history.push('/');

    const { container } = render(<HOC></HOC>);

    const item1 = screen.getByText('todo1');
    const item2 = screen.getByText('todo2');
    const item3 = screen.getByText('todo3');

    expect(item1).toBeInTheDocument();
    expect(item2).toBeInTheDocument();
    expect(item3).toBeInTheDocument();

    const deleteButtons = screen.getAllByText('삭제');
    expect(deleteButtons).toHaveLength(3);

    const addButton = screen.getByText('+');
    expect(addButton).toBeInTheDocument();

    expect(addButton.getAttribute('href')).toBe('/add');
    expect(container).toMatchSnapshot();
  });

  it('Delete Button Test', () => {
    render(<HOC></HOC>);

    const item1 = screen.getByText('todo1');
    expect(item1).toBeInTheDocument();

    const deleteButton = screen.getAllByText('삭제')[0];

    fireEvent.click(deleteButton);

    expect(item1).not.toBeInTheDocument();

    expect(JSON.parse(localStorage.getItem(_todo_key) as string)).toHaveLength(2);
    expect(JSON.parse(localStorage.getItem(_todo_key) as string)).not.toContain('todo1');
  });

  it('moves detail page', () => {
    render(<HOC></HOC>);

    const item1 = screen.getByText('todo1');
    expect(item1).toBeInTheDocument();
    expect(item1.getAttribute('href')).toBe('/detail/0');

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();
    fireEvent.click(item1);
    expect(url.textContent).toBe('/detail/0');
  });

  it('move to add page', () => {
    render(<HOC></HOC>);

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const addButton = screen.getByText('+');
    expect(addButton).toBeInTheDocument();

    fireEvent.click(addButton);

    expect(url.textContent).toBe('/add');
  });
});
