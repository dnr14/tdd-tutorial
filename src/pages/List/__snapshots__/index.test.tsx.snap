// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List> <List> 랜더링 확인 1`] = `
.c4 {
  text-align: center;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
  background-color: #ff1744;
}

.c4:hover {
  background-color: #f01440;
}

.c4:active {
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
}

.c5 {
  color: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 1px solid #bdbdbd;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 10px;
  padding: 10px;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 20px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 {
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
}

.c6 {
  font-size: 20px;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 60px;
  height: 60px;
  border-radius: 30px;
  cursor: pointer;
  background-color: #404ffe;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:hover {
  background-color: #1e40ff;
}

.c6:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

<div>
  <div>
    /
  </div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <a
          class="c3"
          href="/detail/0"
        >
          todo1
        </a>
        <div
          class="c4"
          data-testid="button-container"
        >
          <div
            class="c5"
          >
            삭제
          </div>
        </div>
      </div>
      <div
        class="c2"
      >
        <a
          class="c3"
          href="/detail/1"
        >
          todo2
        </a>
        <div
          class="c4"
          data-testid="button-container"
        >
          <div
            class="c5"
          >
            삭제
          </div>
        </div>
      </div>
      <div
        class="c2"
      >
        <a
          class="c3"
          href="/detail/2"
        >
          todo3
        </a>
        <div
          class="c4"
          data-testid="button-container"
        >
          <div
            class="c5"
          >
            삭제
          </div>
        </div>
      </div>
    </div>
    <a
      class="c6"
      href="/add"
    >
      +
    </a>
  </div>
</div>
`;
